tsdb/engine/tsm1/cache.go

```diff
@@ -209,117 +187,61 @@ func (c *Cache) Keys() []string {
 	return a
 }
 
-func (c *Cache) KeyRange(min, max uint64) []string {
-	var a []string
-	for k, v := range c.store {
-		if v.contains(min, max) {
-			a = append(a, k)
+// Values returns a copy of all values, deduped and sorted, for the given key.
+func (c *Cache) Values(key string) Values {
+	values, needSort := func() (Values, bool) {
+		c.mu.RLock()
+		defer c.mu.RUnlock()
+		e := c.store[key]
+		if e == nil {
+			return nil, false
 		}
-	}
-	sort.Strings(a)
-	return a
-}
 
-// Checkpoint returns the current checkpoint for the cache.
-func (c *Cache) Checkpoint() uint64 {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-	return c.checkpoint
-}
-
-func (c *Cache) CheckpointRange() (uint64, uint64) {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-
-	var min, max uint64
-	for _, v := range c.store {
-		for checkpoint := range v.m {
-			if min == 0 || checkpoint < min {
-				min = checkpoint
-			}
-			if max == 0 || checkpoint > max {
-				max = checkpoint
-			}
+		if e.needSort {
+			return nil, true
 		}
-	}
-	return min, max
-}
 
-// Evict forces the cache to evict.
-func (c *Cache) Evict() {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	c.evict()
-}
-
-// Values returns a copy of all values, deduped and sorted as requested, for the given key.
-func (c *Cache) Values(key string, ascending bool) Values {
-	values := func() Values {
-		c.mu.RLock()
-		defer c.mu.RUnlock()
-		e, ok := c.store[key]
-		if !ok {
-			return nil
-		}
-		return e.clone()
+		return e.values[0:len(e.values)], false
 	}()
-	// Now have copy, so perform dedupe and sort outside of lock, unblocking
-	// writes.
 
-	if values == nil {
-		return nil
-	}
-	return values.Deduplicate(ascending)
-}
+	// the values in the entry require a sort, do so with a write lock so
+	// we can sort once and set everything in order
+	if needSort {
```

Comment:
I don't think it is because I'm grabbing the subslice while I have a lock. Although this is one question I had. If you have:

```go
a := []uint64{32, 6, 7}
// then while you have a lock
b := a[0:len(a)-1]

// can you use b in a goroutine and be sure that some other go routine doing
a = append(a, 89)
// wouldn't cause a race?
```